phantazm-zombies/src/main/java/com/github/phantazmnetwork/zombies/equipment/gun/GunLevel.java

```diff
@@ -5,34 +5,77 @@
 import com.github.phantazmnetwork.zombies.equipment.gun.shoot.ShootTester;
 import com.github.phantazmnetwork.zombies.equipment.gun.shoot.fire.Firer;
 import com.github.phantazmnetwork.zombies.equipment.gun.visual.GunStackMapper;
+import net.kyori.adventure.key.Key;
 import net.minestom.server.item.ItemStack;
 import org.jetbrains.annotations.NotNull;
 
 import java.util.Collection;
 import java.util.Objects;
+import java.util.Set;
 
-public record GunLevel(@NotNull ItemStack stack,
+/**
+ * An individual gun level of a gun.
+ * @param upgrades Suggested upgrade {@link Key}s for next levels
+ * @param stack The {@link ItemStack} for this level
+ * @param stats The gun's {@link GunStats}
+ * @param shootTester The gun's {@link ShootTester}
+ * @param reloadTester The gun's {@link ReloadTester}
+ * @param firer The gun's {@link Firer}
+ * @param startEffects The gun's {@link GunEffect}s that are invoked when the gun level becomes active
+ * @param shootEffects The gun's {@link GunEffect}s that are invoked when the gun is shot
+ * @param reloadEffects The gun's {@link GunEffect}s that are invoked when the gun begins reloading
+ * @param tickEffects The gun's {@link GunEffect}s that are invoked every tick
+ * @param noAmmoEffects The gun's {@link GunEffect}s that are invoked when the gun has no ammo
+ * @param gunStackMappers The gun's {@link GunStackMapper}s that produce the visual {@link ItemStack} representation of the gun
+ */
+public record GunLevel(@NotNull Set<Key> upgrades,
+                       @NotNull ItemStack stack,
                        @NotNull GunStats stats,
                        @NotNull ShootTester shootTester,
                        @NotNull ReloadTester reloadTester,
                        @NotNull Firer firer,
+                       @NotNull Collection<GunEffect> startEffects,
```

Comment:
startEffects is a bit ambiguous, i suggest maybe purchaseEffects, initialEffects or startingEffects

Reply:
activateEffects?